File: Paul/Array/49_groupAnagrams.js
// https://leetcode.com/problems/group-anagrams/

/**
 * @param {string[]} strs
 * @return {string[][]}
 */
const groupAnagrams = function (strs) {
  let map = new Map();

  for (word of strs) {
    let str = word.split('').sort().join('');

    map.get(str) || map.set(str, []);
    map.get(str).push(word);
  }

  let result = [];

  for (let value of map.values()) {
    result.push(value);
  }

  return result;
};

// leetcode solution - counting
const groupAnagrams = function (strs) {
  let res = {};
  for (let str of strs) {
    let count = new Array(26).fill(0);
    for (let char of str) count[char.charCodeAt() - 97]++;
    let key = count.join('#');
    res[key] ? res[key].push(str) : (res[key] = [str]);
  }
  return Object.values(res);
};
// Time Complexity: O(n*k) where n is the size of input array and k is the maximum length of string in input array
// Space Complexity: O(n)

File: P/49_Group_Anagrams.java
//Sorting String (Time: O(NMlogM) SPace: O(NM)) Where N is strs.length and M is strs[i].length
class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        Map<String, List<String>> map = new HashMap<>();
        
        for(int i = 0; i < strs.length; i++) {
            char[] charArr = strs[i].toCharArray();
            Arrays.sort(charArr);
            String sortedStr = String.valueOf(charArr);
            
            if(!map.containsKey(sortedStr))
                map.put(sortedStr, new ArrayList<String>());

            map.get(sortedStr).add(strs[i]);
        }
        return new ArrayList<>(map.values());
    }
}

//Counting String by LeetCode (Time: O(NM) Space: O(NM)) Where N is strs.length and M is strs[i].length
class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        Map<String, List<String>> map = new HashMap<>();
        int[] count = new int[26];
        for(String str: strs) {
            for(char ch: str.toCharArray())
                count[ch - 'a']++;

            StringBuilder sb = new StringBuilder();
            for(int i = 0; i < count.length; i++)
                sb.append(count[i] + "#");

            String codedStr = sb.toString();
            if(!map.containsKey(codedStr))
                map.put(codedStr, new ArrayList<String>());

            map.get(codedStr).add(str);
            Arrays.fill(count, 0);
        }
        return new ArrayList<>(map.values());
    }
}